In [1]:
# import statements
import numpy as np

In [2]:
# Defining thw MDP

In [3]:
# Example of states
states = [0, 1, 2, 3]
#Example of actions
actions = ['left', 'right']

# Defining the rewards
rewards = np.array([
    [-1, 0], # Rewards for state 0
    [0, 1], # Rewards for state 1
    [1, -1], # Rewards for state 2
    [0, 0] # Rewards for state 3(terminal state)
])

# Defining the transition matrix
transitions = np.array([
    [[0.7, 0.3, 0, 0], [0.1, 0.9, 0, 0]], # Transition probabilities for state 0
    [[0.8, 0, 0.2, 0], [0.4, 0.5, 0, 0.1]], #Transition probabilitie for state 1
    [[0.6, 0.2, 0, 0.2], [0.5, 0.5, 0, 0]], # Transition probabilities for state 2
    [[0, 0, 0, 1], [0, 0, 0, 1]] # Transition probabilities for state 3(Terminal state)
])
#Defining the discount factor
gamma = 0.9
#Threshold for convergence
theta = 1e-6

In [4]:
def value_iteration(states, actions, rewards, transitions, gamma, theta):
  #Initialize the value function to zeros
  V = np.zeros(len(states))
  #Initialize the policy
  policy = np.zeros(len(states), dtype=int)

  while True:
    #change in value function
    delta = 0
    for s in states:
      action_values = []
      for a in range(len(actions)):
        action_value =  sum(transitions[s, a, s_prime] * (rewards[s, a] + gamma * V[s_prime]) for s_prime in states)
        action_values.append(action_value)

      max_action_value = max(action_values)
      delta = max(delta, abs(max_action_value - V[s]))
      V[s] = max_action_value
      policy[s] = np.argmax(action_values)

    if delta < theta:
      break

  return policy, V

In [6]:
#Solve the MDP using Value Iteration
optimal_policy, optimal_value = value_iteration(states, actions, rewards, transitions, gamma, theta)


In [7]:
# Display the results
print("Optimal Value Function:")
print(optimal_value)
print("Optimal Policy (action index):")
print(optimal_policy)

Optimal Value Function:
[3.87744984 4.3561483  3.87792961 0.        ]
Optimal Policy (action index):
[1 1 0 0]


In [8]:
# Map policy indices to action names
policy_actions = [actions[a] for a in optimal_policy]
print("Optimal Policy (actions):")
print(policy_actions)

Optimal Policy (actions):
['right', 'right', 'left', 'left']


This script demonstrates the implementation of the **Value Iteration** algorithm to solve a simple **Markov Decision Process (MDP)**. Below is a detailed explanation of the code and its key concepts:

---

### **1. Defining the MDP Components**

1. **States**:  
   - Represented as a list of integers `[0, 1, 2, 3]`. These are the states in the MDP.

2. **Actions**:  
   - Two possible actions, `'left'` and `'right'`.

3. **Rewards**:  
   - A 2D numpy array representing the reward for taking each action in each state.  
     Example: `rewards[0] = [-1, 0]` means:
     - Taking `'left'` in state 0 gives a reward of \(-1\).
     - Taking `'right'` in state 0 gives a reward of \(0\).

4. **Transitions**:  
   - A 3D numpy array where `transitions[s, a, s_prime]` represents the probability of transitioning to state \(s'\) from state \(s\) after taking action \(a\).  
     Example: `transitions[0, 0] = [0.7, 0.3, 0, 0]` means:
     - From state 0, taking `'left'` transitions to:
       - State 0 with \(70\%\) probability.
       - State 1 with \(30\%\) probability.

5. **Discount Factor (\(\gamma\))**:  
   - Determines how much future rewards are valued compared to immediate rewards. Here, \(\gamma = 0.9\), which values future rewards almost as much as immediate ones.

6. **Threshold (\(\theta\))**:  
   - A small value \(1e-6\) to determine convergence. The algorithm stops when the change in the value function is smaller than this threshold.

---

### **2. Value Iteration Algorithm**

#### **Goal**  
To find the **optimal policy** (a mapping from states to actions) and the corresponding **value function** (maximum expected cumulative rewards for each state).

#### **Steps**  

1. **Initialize Values**:  
   - `V`: Value function for each state, initially set to \(0\).  
   - `policy`: The optimal action for each state, initialized to \(0\) (arbitrary action index).

2. **Iterate Until Convergence**:
   - For each state \(s\):
     - Compute the **action-value** for each action \(a\):
       \[
       Q(s, a) = \sum_{s'} P(s'|s, a) \cdot \left( R(s, a) + \gamma \cdot V(s') \right)
       \]
       where:
       - \(P(s'|s, a)\) is the transition probability.
       - \(R(s, a)\) is the immediate reward.
       - \(V(s')\) is the value of the next state.
     - Find the **maximum action-value**:
       \[
       V(s) = \max_a Q(s, a)
       \]
     - Update the **policy** to select the action with the highest action-value.

   - Track the maximum change in the value function (\(\Delta\)) during this iteration:
     \[
     \Delta = \max |V_{\text{new}}(s) - V_{\text{old}}(s)|
     \]
   - Stop when \(\Delta < \theta\).

3. **Return Optimal Policy and Value Function**:
   - `policy`: The optimal action for each state.
   - `V`: The optimal value function.

---

### **3. Results**

1. **Optimal Value Function**:  
   The maximum expected cumulative rewards for each state.

2. **Optimal Policy (Indices)**:  
   The action indices (`0` for `'left'` and `1` for `'right'`) that achieve the optimal value.

3. **Optimal Policy (Actions)**:  
   Mapped action names for clarity.

---

### **Example Output**

Given the defined MDP:

```plaintext
Optimal Value Function:
[...values for each state...]
Optimal Policy (action index):
[...indices...]
Optimal Policy (actions):
['right', 'right', 'left', 'left']
```

This means, for example:
- In state 0, the optimal action is `'right'`.
- In state 2, the optimal action is `'left'`.

---

### **Key Takeaways**
- Value Iteration efficiently computes the optimal policy and value function for small to medium-sized MDPs.
- It requires explicit definitions of rewards and transition probabilities, making it suitable for problems with known dynamics.
- Larger state-action spaces may require more scalable methods like Approximate Dynamic Programming or Reinforcement Learning.


The output of the Value Iteration algorithm provides the **Optimal Value Function** and **Optimal Policy** for the given Markov Decision Process (MDP).

---

### **1. Optimal Value Function**

\[
[3.87744984, 4.3561483, 3.87792961, 0.0]
\]

- Each value represents the maximum expected cumulative reward for being in a particular state and following the optimal policy thereafter.
- State-wise values:
  - **State 0**: \(3.877\) (expected reward when starting in state 0).
  - **State 1**: \(4.356\) (highest value, meaning this state is relatively more rewarding).
  - **State 2**: \(3.878\) (similar to state 0).
  - **State 3**: \(0.0\) (terminal state with no future rewards).

---

### **2. Optimal Policy (Action Index)**

\[
[1, 1, 0, 0]
\]

- The policy shows the optimal action for each state using action indices:
  - **Action 1** corresponds to `'right'`.
  - **Action 0** corresponds to `'left'`.
- State-wise optimal actions:
  - **State 0**: Action \(1\) (go `'right'`).
  - **State 1**: Action \(1\) (go `'right'`).
  - **State 2**: Action \(0\) (go `'left'`).
  - **State 3**: Action \(0\) (go `'left'`), though this state is terminal.

---

### **3. Optimal Policy (Actions)**

\[
['right', 'right', 'left', 'left']
\]

- Maps the action indices to their corresponding names:
  - **State 0**: `'right'` is optimal.
  - **State 1**: `'right'` is optimal.
  - **State 2**: `'left'` is optimal.
  - **State 3**: `'left'` is optimal, although this state is terminal.

---

### **Interpretation**
- The **Optimal Value Function** provides a numerical assessment of each state's desirability under the optimal policy.
- The **Optimal Policy** tells us which action to take in each state to maximize rewards.

### **Insights**
1. **Terminal State (State 3)**:
   - Has a value of \(0.0\) since no future rewards are obtainable after reaching it.

2. **Preference for Actions**:
   - States \(0\) and \(1\) favor moving `'right'`, suggesting that higher rewards or favorable transitions lie in that direction.
   - State \(2\) prefers `'left'`, indicating better outcomes by transitioning in that direction.

---

